File: golang-devildyw/middle/L230二叉搜索树中第K小的元素.go
package main

func main() {}

func kthSmallest(root *TreeNode, k int) int {
	var stack []*TreeNode
	stack = append(stack, pushLeft(root)...)
	result := -1
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		k--
		if k == 0 {
			result = node.Val
			break
		}

		if node.Right != nil {
			stack = append(stack, pushLeft(node.Right)...)
		}
	}
	return result
}

func pushLeft(root *TreeNode) []*TreeNode {
	var stack []*TreeNode
	for root != nil {
		stack = append(stack, root)
		root = root.Left
	}
	return stack
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L230二叉搜索树中第K小的元素.java
package top.dyw.leetcode.Hot100;

public class L230二叉搜索树中第K小的元素 {
    int result = -1;
    int count = 0;
    public int kthSmallest(TreeNode root, int k) {
        if (root == null) {
            return -1;
        }
        kthSmallestHelper(root, k);
        return result;
    }

    public void kthSmallestHelper(TreeNode root, int k) {
        if (root == null) {
            return ;
        }

        if (count > k) {
            return;
        }

        kthSmallestHelper(root.left, k);
        count++;
        if (count==k) {
            result = root.val;
            return ;
        }

        kthSmallestHelper(root.right, k);

    }
}
